htsget-config/examples/config-files/basic.toml

```diff
@@ -0,0 +1,18 @@
+# An example of running htsget-rs.
```

Comment:
We'll need to discuss/decide on what to do with the `bin/settings.ts` and `*.toml` situation, I'm not sure adding yet another toml file helps? :___S

Reply:
The TOML files will always stay around for htsget-rs right? Unless we want to use something like YAML or JSON (which I think are not as nice)?

The only thing that I think will get updated is that whatever is defined in `settings.ts` won't have to be doubly-defined in a TOML file as well. I.e. the toml files under `htgset-config/examples/config-files` aren't related to the ones under `deploy/config`.

Note, that this is only for the CDK deployment,  I don't think we can use `settings.ts` for running htsget-rs locally for example?

Reply:
Sorry for the delay, I wanted to think this through a little bit before answering (haven't forgotten!) ;)

> Note, that this is only for the CDK deployment, I don't think we can use settings.ts for running htsget-rs locally for example?

If it's only for the CDK deployment, how come we don't define ALL config parameters on the Stack itself and remove the possibility/option of touching `settings.ts` and/or any of the `*.toml` files?

I see/understand that we also have the locally running option but in any case, there should be only ONE place to define configuration?

Reply:
> If it's only for the CDK deployment, how come we don't define ALL config parameters on the Stack itself and remove the possibility/option of touching settings.ts and/or any of the *.toml files?

Happy to define it on the stack directly. The `settings.ts` is already just a definition on the [props](https://github.com/umccr/htsget-rs/blob/main/deploy/bin/htsget-lambda.ts#L25) defined on the construct. I placed it in a separate file to make it clearer (rather than nested inside the props), although it would be simple to remove.

I think one of the problems is that we have two kinds of config. We have the htsget-rs server config, things like the address, resolvers configuration, CORS, service info, etc. We also have AWS-specific config, things like the Route53 domain names, s3 bucket permissions, etc. I think for CDK-related deployments, it makes sense to define all config inside the construct [props](https://github.com/umccr/htsget-rs/blob/274bebbf9f49e9c8e03fef77c279f43fc3cb5669/deploy/lib/htsget-lambda-stack.ts#L51-L128) (whether this is in a separate `settings.ts` or not). CDK is the language we use to deploy AWS, so we should probably use it's configuration patterns. This means that all of the htsget-rs server config should go inside [`HtsgetSettings`](https://github.com/umccr/htsget-rs/blob/274bebbf9f49e9c8e03fef77c279f43fc3cb5669/deploy/lib/htsget-lambda-stack.ts#L51-L128). For example, there could be another type that defines service info:

```ts
export type ServiceInfo = {
    ...
    organization_name: string;
    organization_url: string;
    ...
}
````

> If it's only for the CDK deployment, how come we don't define ALL config parameters on the Stack itself and remove the possibility/option of touching settings.ts and/or any of the *.toml files?

These props would then be converted to env variables to be consumed by the Lambda function. This means that for CDK, there will be just one place to define config (inside props). I think this configuration wouldn't support every option that the TOML files support. For example, there's no point to allow a `Local` resolver, because the server is being deployed as a Lambda function.

Reply:
The htsget-rs could also be deployed in a way that doesn't use CDK (not everyone's using AWS), such as a `Dockerfile`. In this case, I don't think it makes sense for the server to be configured using typescript props files, as it's a Rust program?

I also don't think it makes sense to have the configuration be defined only by env variables or on the command line. For example, because resolvers is an array of struct definitions, it's not very easy to use [env](https://github.com/umccr/htsget-rs/blob/main/htsget-config/README.md#environment-variables) variables. I don't think this would be any simpler on the command line either?

This leaves the option of using configuration files, whether than toml, json, yaml, etc. Currently, for the htsget-rs server (non-CDK deployment), this is the single way that can define all config parameters. I think it's okay this is way is different to the CDK deployment props?

There is a question of how to make the config simpler, maybe if it's simplified, it could be configured using just env variables, or on the command line. It's not clear to me to how to simplify things though, as there's a lot of different options. E.g. consider the three different resolver types, S3, Local, and Url - and now we also have C4GH. 

The other thing to note is that I think it's okay for the user to have multiple ways to configure something. For example, the htsget-rs server supports either env or config files. There's no requirement to use config files if you don't want to (although I think it makes sense for resolvers).

Reply:
Here's a thought on how to simplify config: 

* It looks like most of the time, when using resolvers, there is just one storage type for a deployment. For example, the Lambda deployment always uses the S3 resolver for us. Instead of defining the resolvers as an array of storage types, if the user specified a flattened config for the resolvers, the server could assume that the user means that they want a single storage type for all resolvers?

Happy for suggestions here! I think it's okay if the user can opt-in to more complex configuration, as long as simple cases are easy to use?

Reply:
That last trick looks neat! I'd also target the regex-based config since it's a common source of confusion/frustration and I'm not sure if it has seen much use "in the wild": users just define a bucket's name and they want all files contained in it accessible?

Still thinking through about our current options laid out here (and re-reading code), thanks for the counterpoints and discussion!

Reply:
> I'd also target the regex-based config since it's a common source of confusion/frustration and I'm not sure if it has seen much use "in the wild": users just define a bucket's name and they want all files contained in it accessible?

I think a similar logic could apply to the regex. Most of the time, the user want's to be able to type `/reads/<id>` and get it to map directly to the corresponding storage location at `<id>.bam` and `<id>.bam.bai`. This corresponds to a regex like:

```
regex = '.*'
substitution_string = '$0'
```

Just to note, this is already the case! As it's the default values for the regex. But maybe the docs needs to be clearer that the user doesn't have to bother defining the regex most of the time.

Reply:
Yes! Reasonable (documented) defaults! Secure conventions over configuration!

Reply:
Do you think there is any way to get rid of the regex altogether? I feel like if the user want's to use different kinds of storage backends (e.g. some files are C4GH, some are S3, some are URL), then regex is probably the easiest way to express this logic. Although in the default case this probably should be an opt-in feature.

Reply:
> then regex is probably the easiest way to express this logic

Is it? Wouldn't static strings/paths be easier on the user's cognitive's load (and code detecting S3/HTTP/FILE schemes under the hood)? Or even futher: imposing a convention that the user has to comply with, killing any required config?

Reply:
How would you define a scenario like:
* The user wants to serve some files from their S3 bucket, and some files from an external server, based on the <id> in `/reads/<id>`? For example, `/reads/prefix1/some_file` needs to fetch data from S3, and `/reads/prefix2/some_file` needs to fetch data from an external server?

Maybe like?
```
location = s3://bucket
prefix = prefix1

location = https://example
prefix = prefix2
```

I think this would work for simpler cases. Is there ever a scenario where something more complex is required? For example, what if the user wants to be able serve different file types from different buckets? Maybe in this case, an opt-in to regex makes sense (although it should probably be under "advanced configuration")?

Reply:
> How would you define a scenario like: (...)

I'd probably go for sth like this (assuming everything under the prefixes should be accessible, which is pretty close to 100% of our use cases so far?):

```
location = s3://bucket/prefix1
location = https://example/prefix2
```

That's why I have a bias towards env vars/cmdline options... having a file to just define those two lines is overkill, IMHO.

> although it should probably be under "advanced configuration"

Yes to regex usecases tucked under "advanced config"... I'm not against advanced "escape hatches", I'm just saying that there's a chance to cater to users like GA4GH via simplified defaults/conventions (as we originally plotted in https://github.com/umccr/htsget-rs/issues/269#issue-2577624756 last 2 points)... or without even using config files (for the most common, 90% of use cases).

Reply:
Fair enough! I think that one-liner is probably okay but not sure.

Is there any chance that leads to confusion? There's kind of an implicit logic that says that everything after the first first '/' (not including `://`) is what the <id> has to be prefixed with for it to go to that endpoint.

Reply:
No, I think that (implicit) assumption is very reasonable, something users should expect when seeing example `curl` commands in the docs?